### PR TITLE
Add cycles & project milestones + automated testing

### LINKED ISSUE #1: `project.targetDate?.toISOString` when running `linearis projects list`
There seems to be a bug here. Pssobile due to the Node version I have installed or something?

`/opt/homebrew/Cellar/node/24.10.0/bin/node`

```sh
$ linearis -V
1.0.0
$ linearis projects list
{
  "error": "project.targetDate?.toISOString is not a function"
}
$ linearis issues list
[
  {
     ...
  }
]
```

### LINKED ISSUE #2: Support for Cycles in Linearis CLI
It looks like the Linearis CLI currently supports querying issues and projects, but not cycles. It would be really helpful to have cycle support as well — for example, being able to:

- List all cycles
- Retrieve the current cycle
- Add or remove issues from a cycle
- Query for issues within a specific cycle

This would make it easier to do planning and reporting directly from the LLM  without needing to switch back to the Linear app.

### LINKED ISSUE #3: Feature Request: Add milestone support to issues update command
## Feature Request

The `issues update` command currently does not support updating milestones, while the `issues create` command does support the `--milestone` option.

### Current Behavior

```bash
linearis issues create "Title" --project "Project" --milestone "Milestone" --team "Team"  # ✅ Works
linearis issues update ISSUE-123 --milestone "Milestone"  # ❌ Not supported
```

### Requested Behavior

Add `--milestone` option to the `issues update` command to allow changing an issue's milestone after creation.

```bash
linearis issues update ISSUE-123 --milestone "New Milestone"
linearis issues update ISSUE-123 --clear-milestone  # Optional: clear milestone
```

### Use Case

When creating issues programmatically or in batch, it's often useful to assign milestones after the fact, especially when:
- Issues are created without knowing the milestone initially
- Milestones need to be changed during project planning
- Automating issue management workflows

### Implementation Notes

Similar to how `--parent-ticket` has both a setter and a clearer:
- `--milestone <milestone>` - set or change milestone (by name or ID)
- `--clear-milestone` - remove milestone assignment

This would be consistent with the existing `issues create` milestone support and follow the same pattern as other update options.

### Workaround

Currently, the only workaround is to update milestones manually in the Linear UI, which is not ideal for automation.

Thank you for maintaining this excellent CLI tool!

### PR DESCRIPTION
**Pull request by ryanrozich**
# Add Cycles & Project Milestones + Automated Testing

## Summary

This PR adds comprehensive cycles and project milestones support to linearis, fixes two critical bugs, and establishes automated testing infrastructure.

### Features Added
- **Cycles Management**: List, read, create, update, and delete cycles with smart filtering
- **Project Milestones**: Full CRUD operations for project milestones
- **Smart ID Resolution**: Resolve cycles by name or ID with automatic disambiguation

### Bug Fixes
- **Command Naming**: Renamed `projectMilestones` to `project-milestones` for CLI consistency
- **GraphQL Complexity**: Refactored cycles to use SDK pagination, eliminating "query too complex" errors

### Testing Infrastructure
- **27 automated tests** (14 unit, 13 integration) - all passing
- **35% command coverage** (cycles and project-milestones fully covered)
- **GitHub Actions CI/CD pipeline**
- **Custom command coverage tool** for CLI-specific metrics

### Closes
- Closes #1 (bug: `project.targetDate` error)
- Closes #2 (feature: cycles support)
- Closes #3 (feature: milestone support)

### List Cycles
```bash
# List all cycles
linearis cycles list

# Filter by team
linearis cycles list --team Engineering

# Show only active cycles
linearis cycles list --active --team Backend

# Show cycles around the active one (±N cycles)
linearis cycles list --around-active 5 --team Backend
```

### Read Cycle
```bash
# By ID
linearis cycles read cycle-uuid

# By name (with team disambiguation)
linearis cycles read "Sprint 2025-11" --team Backend

# Limit number of issues returned
linearis cycles read cycle-uuid --issues-first 100
```

**Features**:
- ✅ Smart name resolution with disambiguation (prefers active → next → previous)
- ✅ Automatic pagination (handles 500+ cycles without complexity errors)
- ✅ Rich cycle data including progress, issue counts, and team info

### List Milestones
```bash
# List milestones for a project
linearis project-milestones list --project "Mobile App"

# Limit results
linearis project-milestones list --project "Backend API" -l 10
```

### Read Milestone
```bash
# By ID
linearis project-milestones read milestone-uuid

# By name (with project disambiguation)
linearis project-milestones read "MVP Release" --project "Mobile App"

# Include issues
linearis project-milestones read milestone-uuid --issues-first 50
```

### Create Milestone
```bash
linearis project-milestones create "Q1 Goals" \
  --project "Mobile App" \
  --description "First quarter objectives" \
  --target-date 2025-03-31
```

### Update Milestone
```bash
linearis project-milestones update "Q1 Goals" \
  --project "Mobile App" \
  --target-date 2025-04-15 \
  --description "Updated timeline"
```

**Features**:
- ✅ Project-scoped name resolution
- ✅ Target date tracking
- ✅ Issue association and progress tracking

### Issue 1: Command Naming Inconsistency

**Before**: `linearis projectMilestones` (camelCase - didn't work as expected)
**After**: `linearis project-milestones` (kebab-case - works correctly)

**Impact**: Users can now discover and use the command with standard CLI conventions.

### Issue 2: GraphQL Complexity Errors

**Before**: Direct GraphQL queries failed with "query too complex" for workspaces with 500+ cycles
**After**: SDK-based pagination handles any workspace size

**Changes**:
- Added `getCycles()`, `getCycleById()`, `resolveCycleId()` to LinearService
- Refactored cycles.ts from 135 lines to 78 lines
- Uses Linear SDK's automatic pagination (250 cycles per request)
- Maintains all functionality (--team, --active, --around-active)

### Test Suite
```bash
# Run all tests
pnpm test                    # 27 tests in ~15 seconds

# Development workflow
pnpm test:watch              # Watch mode
pnpm test:ui                 # Interactive UI

# Coverage reports
pnpm test:coverage           # Code coverage
pnpm test:commands           # Command coverage
```

### What's Tested

**Unit Tests** (14 tests):
- LinearService cycle methods with mocks
- Error handling and edge cases
- Date conversion and filtering
- Cycle name disambiguation logic

**Integration Tests** (13 tests):
- End-to-end CLI functionality
- All command flags and options
- JSON output structure
- No "query too complex" errors verified
- Auto-skips without LINEAR_API_TOKEN

**Command Coverage**: 35% (7/20 commands)
- ✅ cycles list, cycles read
- ✅ project-milestones list
- ✅ projects list

### CI/CD Pipeline

GitHub Actions workflow runs on every push:
- ✅ Build verification
- ✅ TypeScript type checking
- ✅ Test suite (unit + integration)
- ✅ Automated on all PRs

### Automated Checks ✅
- [x] Build passes: `pnpm run build`
- [x] All 27 tests pass: `pnpm test`
- [x] TypeScript compiles cleanly
- [x] Command coverage: 35%

### Manual Testing

**Cycles Commands**:
```bash
# Should work without complexity errors
linearis cycles list --team <YOUR_TEAM>
linearis cycles list --around-active 5 --team <YOUR_TEAM>
linearis cycles read <CYCLE_ID>
```

**Project Milestones Commands**:
```bash
# Should use kebab-case naming
linearis project-milestones --help
linearis project-milestones list --project "<PROJECT_NAME>"
```

**Regression Tests**:
```bash
# Other commands should still work
linearis issues list -l 5
linearis projects list
```

### Architecture

**Cycles Implementation**:
```
Before: cycles command → rawRequest() → Single GraphQL query
After:  cycles command → LinearService → SDK with pagination
```

**Files Changed**:
- `src/commands/cycles.ts` - New cycles command (78 lines)
- `src/commands/projectMilestones.ts` - New milestones command (308 lines)
- `src/utils/linear-service.ts` - Add cycle methods (+169 lines)
- `src/queries/cycles.ts` - GraphQL queries (deprecated, preserved for reference)
- `src/queries/projectMilestones.ts` - GraphQL queries (149 lines)

**Testing Files**:
- `tests/unit/linear-service-cycles.test.ts` - 14 unit tests
- `tests/integration/cycles-cli.test.ts` - 8 integration tests
- `tests/integration/project-milestones-cli.test.ts` - 5 integration tests
- `tests/command-coverage.ts` - Command coverage analyzer
- `.github/workflows/ci.yml` - CI/CD pipeline

### Smart ID Resolution

**Cycles**:
- Accepts UUID or cycle name
- Disambiguates multiple matches: active → next → previous
- Team filtering for disambiguation
- Clear error messages for ambiguous names

**Project Milestones**:
- Accepts UUID or milestone name
- Project-scoped name lookup
- Falls back to global search if needed
- Disambiguation with helpful error messages

## Breaking Changes

**None** - All changes are backward compatible:
- New commands added (cycles, project-milestones)
- Existing commands unchanged
- Command naming fix makes previously broken command work

## Performance

**Cycles with Large Datasets**:
- Before: Failed at ~75+ cycles (complexity error)
- After: Handles 500+ cycles reliably
- Trade-off: ~100-200ms slower for small datasets (acceptable)

## Future Work

**Testing Coverage Goals**:
- [ ] Issues commands (list, read, create, update, search)
- [ ] Labels list
- [ ] Comments create
- [ ] Project milestones create/update/read

**Target**: 50%+ command coverage

## Files Changed

**Source Code**: 32 files
- Core features: 7 files
- Testing: 5 files
- CI/CD: 1 file
- Documentation: 3 files
- Compiled output: 16 files

**Line Changes**: +4,958 / -196

## Documentation

- `docs/testing.md` - Comprehensive testing guide
- `CLAUDE.md` - Updated with testing framework
- `.gitignore` - Exclude coverage and test files

## Checklist

- [x] Features work as documented
- [x] Bug fixes verified by tests
- [x] All automated tests pass (27/27)
- [x] No TypeScript errors
- [x] Documentation updated
- [x] CI/CD pipeline configured
- [x] Backward compatible
- [x] No regressions detected

---

**Ready for Review**: All features implemented, bugs fixed, and thoroughly tested with 27 automated tests.

### CODE REVIEW

File: src/utils/graphql-issues-service.ts [hunk 1]
```diff
@@ -245,10 +329,10 @@ export class GraphQLIssuesService {
       updateInput.assigneeId = args.assigneeId;
     }
     if (finalProjectId !== undefined) updateInput.projectId = finalProjectId;
+  if (finalCycleId !== undefined) updateInput.cycleId = finalCycleId;
```
Comment (czottmann) [inline src/utils/graphql-issues-service.ts]: Minor indentation issue, 2 spaces instead of 4

File: src/commands/projectMilestones.ts [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    .action(
+      handleAsyncCommand(
+        async (milestoneIdOrName: string, options: any, command: Command) => {
+          const graphQLService = await createGraphQLService(
+            command.parent!.parent!.opts(),
+          );
+
+          let milestoneId = milestoneIdOrName;
+
+          if (!isUuid(milestoneIdOrName)) {
```
Comment (czottmann) [inline src/commands/projectMilestones.ts]: This milestone resolution logic appears both in the "read" and "update" commands, and is nearly identical code. I'd put it in a helper function.

File: dist/commands/projectMilestones.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        .description("Get milestone details including issues. Accepts UUID or milestone name (optionally scoped by --project)")
+        .option("--project <project>", "project name or ID to scope name lookup")
+        .option("--issues-first <n>", "how many issues to fetch (default 50)", "50")
+        .action(handleAsyncCommand(async (milestoneIdOrName, options, command) => {
+        const graphQLService = await createGraphQLService(command.parent.parent.opts());
+        let milestoneId = milestoneIdOrName;
+        if (!isUuid(milestoneIdOrName)) {
+            let nodes = [];
+            if (options.project) {
+                let projectId = options.project;
```
[the remark this answers is not in the document]
Comment (ryanrozich) [reply dist/commands/projectMilestones.js]: Fixed - extracted resolveProjectId() helper function

File: src/commands/projectMilestones.ts (hunk 2 above)
Comment (ryanrozich) [reply src/commands/projectMilestones.ts]: Fixed - extracted resolveMilestoneId() helper function

File: src/utils/linear-service.ts [hunk 4]
```diff
@@ -333,6 +333,175 @@ export class LinearService {
...
+    }
+
+    if (activeOnly) {
+      filter.isActive = { eq: true };
+    }
+
+    const cyclesConnection = await this.client.cycles({
+      filter: Object.keys(filter).length > 0 ? filter : undefined,
+      orderBy: "createdAt" as any,
+      first: 250,
```
[the remark this answers is not in the document]
Comment (ryanrozich) [reply src/utils/linear-service.ts]: Fixed - extracted as DEFAULT_CYCLE_PAGINATION_LIMIT constant

File: src/queries/cycles.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
+/**
```
[the remark this answers is not in the document]
Comment (ryanrozich) [reply src/queries/cycles.ts]: You're right - will remove these deprecated queries since they're not being used. The SDK-based approach replaced them entirely.

[the remark this answers is not in the document]
Comment (czottmann) [reply src/queries/cycles.ts]: They are still here…?

**Comment by chadrwalters** [comment]
@claude, @bugbot, @codex, @gemini — please review

bugbot run

## Review Request for #7: Add cycles & project milestones + automated testing

**Branch**: `pr-4` → `release/2025.11.2`

## Summary Since Last Push
Files changed: 21
Commits: 13
Total changes: 5390 (extra_large)

Recent commits:
- feat(types): add project label and CRUD input types — Chad Walters
- feat(projects): add createProject method to LinearService — Chad Walters
- feat(projects): add updateProject, getProjectById, archive/unarchive … — Chad Walters
- feat(projects): add read, create, update, archive, unarchive commands — Chad Walters
- feat(milestones): add delete command for project milestones — Chad Walters

## Focus Areas
**Risk**: medium (Large change set; PR marked as draft)
High-churn files:
- docs/plans/2026-01-20-project-label-crud.md (+2981/-0)
- src/utils/linear-service.ts (+436/-0)
- tests/unit/commands/projects.test.ts (+271/-2)
Key review tasks:
- github-actions: ## 🏛️ Forseti Score Report - **Score:** 83.0
**Tier:** ❓ silver (Goal: GOLD)
`████████████████████` 83.
- github-actions: ## 📊 Coverage Report - | Metric | Current | Main | Delta | Goal |
|--------|---------|------|-------|------|
| **Overall** | 78.3% | N/A | N/A | 80% |
**Progress to 80%:**
████████████████████ 78.3%
**Status:** 📝 First run - baseline recorded

Ready for review.
